File: app/src/main/java/com/sharecloud/sharefiles/sharedata/anyshare/musictransfer/filetransfer/ui/fragment/ImageListFragment.java
package com.sharecloud.sharefiles.sharedata.anyshare.musictransfer.filetransfer.ui.fragment;

/**
 * Created by gabm on 11/01/18.
 */

import android.content.Context;
import android.os.Bundle;
import android.provider.MediaStore;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;

import com.sharecloud.sharefiles.sharedata.anyshare.musictransfer.filetransfer.R;
import com.sharecloud.sharefiles.sharedata.anyshare.musictransfer.filetransfer.callback.Callback;
import com.sharecloud.sharefiles.sharedata.anyshare.musictransfer.filetransfer.app.GalleryGroupEditableListFragment;
import com.sharecloud.sharefiles.sharedata.anyshare.musictransfer.filetransfer.ui.adapter.ImageListAdapter;
import com.sharecloud.sharefiles.sharedata.anyshare.musictransfer.filetransfer.ui.callback.TitleSupport;
import com.sharecloud.sharefiles.sharedata.anyshare.musictransfer.filetransfer.util.AppUtils;
import com.sharecloud.sharefiles.sharedata.anyshare.musictransfer.filetransfer.widget.GroupEditableListAdapter;

import org.jetbrains.annotations.NotNull;

public class ImageListFragment
        extends GalleryGroupEditableListFragment<ImageListAdapter.ImageHolder, GroupEditableListAdapter.GroupViewHolder, ImageListAdapter>
        implements TitleSupport {
    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setFilteringSupported(true);
        setDefaultOrderingCriteria(ImageListAdapter.MODE_SORT_ORDER_DESCENDING);
        setDefaultSortingCriteria(ImageListAdapter.MODE_SORT_BY_DATE);
        setDefaultViewingGridSize(3, 4);
        setUseDefaultPaddingDecoration(false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        setEmptyImage(R.drawable.ic_photo_white_24dp);
        setEmptyText(getString(R.string.text_listEmptyImage));
    }

    @Override
    public void onResume() {
        super.onResume();

        getContext().getContentResolver()
                .registerContentObserver(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, true, getDefaultContentObserver());
    }

    @Override
    public void onPause() {
        super.onPause();

        getContext().getContentResolver()
                .unregisterContentObserver(getDefaultContentObserver());
    }

    @Override
    public ImageListAdapter onAdapter() {
        final AppUtils.QuickActions<GroupEditableListAdapter.GroupViewHolder> quickActions =
                new AppUtils.QuickActions<GroupEditableListAdapter.GroupViewHolder>() {
            @Override
            public void onQuickActions(final GroupEditableListAdapter.GroupViewHolder clazz) {
                if (!clazz.isRepresentative()) {
                    registerLayoutViewClicks(clazz);

                    View visitView = clazz.getView().findViewById(R.id.visitView);
                    visitView.setOnClickListener(
                            new View.OnClickListener() {
                                @Override
                                public void onClick(View v) {
                                    if (getSelectionConnection() != null) {
                                        getSelectionConnection().setSelected(clazz.getAdapterPosition());
                                        Callback.setColor(true);
                                    } else{
                                        performLayoutClick(clazz);
                                        Callback.setColor(false);
                                    }

                                }
                            });

                    visitView.setOnLongClickListener(new View.OnLongClickListener() {
                        @Override
                        public boolean onLongClick(View v) {
                            performLayoutClickOpen(clazz);
                            return true;
                        }
                    });

                    clazz.getView().findViewById(getAdapter().isGridLayoutRequested()
                            ? R.id.selectorContainer : R.id.selector)
                            .setOnClickListener(new View.OnClickListener() {
                                @Override
                                public void onClick(View v) {
                                    if (getSelectionConnection() != null) {
                                        getSelectionConnection().setSelected(clazz.getAdapterPosition());
                                        Callback.setColor(true);
                                    } else
                                        Callback.setColor(false);
                                }
                            });
                }
            }
        };

        return new ImageListAdapter(getActivity()) {
            @NonNull
            @Override
            public GroupViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {

                GroupViewHolder holder = super.onCreateViewHolder(parent, viewType);

                if (viewType == GroupEditableListAdapter.VIEW_TYPE_ADS_GRID) {
                    registerLayoutViewClicks(holder);
                    return holder;
                }

                return AppUtils.quickAction(super.onCreateViewHolder(parent, viewType), quickActions);
            }
        };
    }

    @Override
    public boolean onDefaultClickAction(GroupEditableListAdapter.GroupViewHolder holder) {
        if (getSelectionConnection() != null) {
            Callback.setColor(true);
            return getSelectionConnection().setSelected(holder);
        } else {
            Callback.setColor(false);
            return performLayoutClickOpen(holder);
        }
    }

    @NotNull
    @Override
    public CharSequence getTitle(Context context) {
        return context.getString(R.string.text_photo);
    }

    @Override
    protected RecyclerView onListView(View mainContainer, ViewGroup listViewContainer) {

        View adaptedView = getLayoutInflater().inflate(R.layout.fragment_image_list, null, false);
        listViewContainer.addView(adaptedView);

        return super.onListView(mainContainer, (ViewGroup) adaptedView.findViewById(R.id.imagesList));
    }
}

File: app/src/main/java/com/hazelmobile/filetransfer/ui/fragment/ImageListFragment.kt
package com.hazelmobile.filetransfer.ui.fragment


import android.content.Context
import android.net.Uri
import android.os.Bundle
import android.provider.MediaStore
import androidx.fragment.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.GridLayoutManager

import com.hazelmobile.filetransfer.R
import com.hazelmobile.filetransfer.model.ImageHolder
import com.hazelmobile.filetransfer.ui.adapter.ImageListAdapter
import com.hazelmobile.filetransfer.util.TimeUtils
import com.hazelmobile.filetransfer.util.callback.TitleSupport
import kotlinx.android.synthetic.main.fragment_image_list.*


class ImageListFragment : Fragment(), TitleSupport {

    override fun getTitle(context: Context): CharSequence {
        return "Photos"
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(R.layout.fragment_image_list, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val imageList = ArrayList<ImageHolder>()

        val cursor =
            context?.contentResolver?.query(
                MediaStore.Images.Media.EXTERNAL_CONTENT_URI,
                null,
                null,
                null,
                null
            )

        if (cursor != null) {
            if (cursor.moveToFirst()) {

                val idIndex = cursor.getColumnIndex(MediaStore.Images.Media._ID)
                val titleIndex = cursor.getColumnIndex(MediaStore.Images.Media.TITLE)
                val displayIndex = cursor.getColumnIndex(MediaStore.Images.Media.DISPLAY_NAME)
                val albumIndex =
                    cursor.getColumnIndex(MediaStore.Images.Media.BUCKET_DISPLAY_NAME)
                val dateAddedIndex = cursor.getColumnIndex(MediaStore.Images.Media.DATE_ADDED)
                val sizeIndex = cursor.getColumnIndex(MediaStore.Images.Media.SIZE)
                val typeIndex = cursor.getColumnIndex(MediaStore.Images.Media.MIME_TYPE)

                do {
                    val holder = ImageHolder(

                        TimeUtils.formatDateTime(
                            context!!,
                            cursor.getLong(dateAddedIndex) * 1000
                        ).toString(),
                        cursor.getString(albumIndex),
                        Uri.parse(
                            "${MediaStore.Images.Media.EXTERNAL_CONTENT_URI}/${cursor.getInt(
                                idIndex
                            )}"
                        )

                    )

                    imageList.add(holder)

                } while (cursor.moveToNext())
            }

            cursor.close()

            imageView.apply {
                imageView.layoutManager = GridLayoutManager(context, 3)
                val imageListAdapter = ImageListAdapter(imageList, context)
                imageView.adapter = imageListAdapter
            }

        }


    }
}
